### DISCUSSION TITLE
getSession: [CLIENT_FETCH_ERROR] fetch is not defined

### DISCUSSION THREAD

**Discussion by kian-karazmoudeh** [nextauthjs/next-auth#4125] [Help]
im testing next.js api routes with jest and protecting those routes with getSession. I am also mocking the request and response objects with the _node-mocks-http_ package. However when i run the test i get this error:

`[next-auth][error][CLIENT_FETCH_ERROR] https://next-auth.js.org/errors#client_fetch_error fetch is not defined`

I've also set the NEXTAUTH_URL property in the .env file.
also when i test the routes with postman they work fine. my code:

```
export default async (req, res) => {
  const session = await getSession({ req });
  console.log(req);
  if (!session) {
    res.status(401).send("Unauthorized");
    return;
  }

  if (req.method === "GET") {
    await GetServices(req, res, session);
  } else if (req.method === "POST") {
    await CreateService(req, res, session);
  } else res.status(405).send("Invalid Method");
};

```

any suggestions?

**Comment by ThangHuuVu** [comment]
Seems like a jest setup issue to me. Can you post how you configure your tests?

**Reply by kian-karazmoudeh** [reply]
jest.config.json:

```
const nextJest = require("next/jest");

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: "./",
});

// Add any custom config to be passed to Jest
const customJestConfig = {
  // Add more setup options before each test is run
  // setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
  // if using TypeScript with a baseUrl set to the root directory then you need the below for alias' to work
  // moduleDirectories: ["node_modules", "<rootDir>/"],
  // testEnvironment: "jest-environment-jsdom",
  moduleNameMapper: {
    "^@/lib/(.*)$": "<rootDir>/lib/$1",
  },
};

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
module.exports = createJestConfig(customJestConfig);

```

test.js:

```
import { createMocks } from "node-mocks-http";
import createService from "../pages/api/service.js";

test("Should successfully create a new service", async () => {

  const { req, res } = createMocks({
    body: {
      ....
    },
    headers: {
      cookie:
        "next-auth.session-token=.......",
    },
    method: "POST",
  });
  
  await createService(req, res);
  expect(res._getStatusCode()).toBe(400);
});
```

**Reply by jw-miaem** [reply]
did you ever figure this out?

**Reply by kian-karazmoudeh** [reply]
No unfortunately

**Comment by surajsharma** [comment]
i'm on the same error, this is not a jest issue methinks, here's where my `getSession` fails with this error:

```
  const server = new ApolloServer({
    schema,

    context: async ({ req, res }) => {
      // make context available to gql resolver

      const session = await getSession({ req });
      console.log("🚀 ~ file: index.ts ~ line 37 ~ context ~ session", session)

      return { session };
    },
    csrfPrevention: true,
    cache: 'bounded',
    plugins: [
      ApolloServerPluginDrainHttpServer({ httpServer }),
      ApolloServerPluginLandingPageLocalDefault({ embed: true }),
    ],
  });

```

**Reply by mirulkhanal** [reply]
> i'm on the same error, this is not a jest issue methinks, here's where my `getSession` fails with this error:
> 
> ```
>   const server = new ApolloServer({
>     schema,
> 
>     context: async ({ req, res }) => {
>       // make context available to gql resolver
> 
>       const session = await getSession({ req });
>       console.log("🚀 ~ file: index.ts ~ line 37 ~ context ~ session", session)
> 
>       return { session };
>     },
>     csrfPrevention: true,
>     cache: 'bounded',
>     plugins: [
>       ApolloServerPluginDrainHttpServer({ httpServer }),
>       ApolloServerPluginLandingPageLocalDefault({ embed: true }),
>     ],
>   });
> ```

did you ever find a solution to that? I am working on older version of Apollo server with next-auth@4. I am trying to acces next-auth in the server and pass it as a context but it displays the same error 

```
    name: 'ReferenceError'
  },
  url: 'http://localhost:3000/api/auth/session',
  message: 'fetch is not defined'
}
```

**Reply by DeviousLab** [reply]
> did you ever find a solution to that? I am working on older version of Apollo server with next-auth@4. I am trying to acces next-auth in the server and pass it as a context but it displays the same error
> 
> ```
>     name: 'ReferenceError'
>   },
>   url: 'http://localhost:3000/api/auth/session',
>   message: 'fetch is not defined'
> }
> ```
I also faced a similar issue:
```
[next-auth][error][CLIENT_FETCH_ERROR] 
https://next-auth.js.org/errors#client_fetch_error fetch is not defined {
  error: {
    message: 'fetch is not defined',
    stack: 'ReferenceError: fetch is not defined\n' +
     ...
    name: 'ReferenceError'
  },
  url: 'http://localhost:3000/api/auth/session',
  message: 'fetch is not defined'
}
```
I'm not sure if this will solve your issue but according to this [article](https://bobbyhadz.com/blog/javascript-referenceerror-fetch-is-not-defined) the issue might be related to the node version you're running. The global fetch variable is available in [Node.js starting at version 18](https://nodejs.org/de/blog/announcements/v18-release-announce/), and sure enough when I checked my Node version I was still running v16.16.0. When I updated to the latest LTS of v18.13.0, the issue was solved and I faced no more error messages.

**Comment by SashenJayathilaka** [comment]
Update your **Node js Latest Current Version**. After that I think this issue fix.

**Reply by MrJadaml** [reply]
Hmm, I'm on Node 18.17.1, I've also added global fetch to my jest config, but still seeing the fetch error when running tests. NextAuth works fine in the browser.

```
// jest.config
import nextJest from 'next/jest.js'
 
const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})
 
const config = {
  setupFilesAfterEnv: [
    '<rootDir>/jest.setup.ts',
    '<rootDir>/test/setup-env.ts',
  ],
  testEnvironment: 'jest-environment-jsdom',
  globals: { fetch },   👈 
}

export default createJestConfig(config)
```

Test output:
```
Node.js v18.17.1  👈 
  console.error
    [next-auth][error][CLIENT_FETCH_ERROR] 
    https://next-auth.js.org/errors#client_fetch_error fetch is not defined {
      error: {
        message: 'fetch is not defined',
        stack: 'ReferenceError: fetch is not defined
```

fwiw, since this question arose in other issues around this topic:   `NEXTAUTH_URL="http://localhost:3000"`

**Reply by supertgo** [reply]
@MrJadaml Did you find a solution? Same error here.

**Comment by esemeniuc** [comment]
Upgrading from node 16 to node 18 solved it for me
